src/VirtoCommerce.ElasticAppSearch.Data/Services/Builders/SearchFacetsQueryBuilder.cs

```diff
@@ -0,0 +1,244 @@
+    {
+        _fieldNameConverter = fieldNameConverter;
+        _searchFiltersBuilder = searchFiltersBuilder;
+    }
+
+    public IList<FacetRequest> GetFacetRequests(IEnumerable<AggregationRequest> aggregations, Schema schema)
+    {
+        var results = new List<FacetRequest>();
+
+        if (aggregations.IsNullOrEmpty())
+        {
+            return results;
+        }
+
+        aggregations = PreProcessAggregations(aggregations);
+
+        foreach (var aggregation in aggregations)
+        {
+            Facet facet = null;
+
+            switch (aggregation)
+            {
+                case TermAggregationRequest termAggregationRequest:
+                    facet = AddTermAggregationRequest(termAggregationRequest, schema);
+                    break;
+                case RangeAggregationRequest rangeAggregationRequest:
+                    facet = AddRangeAggregationRequest(rangeAggregationRequest, schema);
+                    break;
+            }
+
+            if (aggregation.Filter != null || facet != null)
+            {
+                var facetRequest = new FacetRequest
+                {
+                    Id = aggregation.Id,
+                    FieldName = aggregation.FieldName,
+                    FacetFieldName = facet?.Name,
+                    Filter = _searchFiltersBuilder.ToFilters(aggregation.Filter, schema),
+                    Facet = facet,
+                    FilterName = aggregation.Filter.ToString(),
+                };
+
+                results.Add(facetRequest);
+            }
+        }
+
+        return results;
+    }
+
+    protected virtual IEnumerable<AggregationRequest> PreProcessAggregations(IEnumerable<AggregationRequest> aggregations)
+    {
+        return PrepareFacets(aggregations);
+    }
+
+    protected virtual Facet AddTermAggregationRequest(TermAggregationRequest termAggregationRequest, Schema schema)
+    {
+        if (string.IsNullOrEmpty(termAggregationRequest.FieldName))
+        {
+            return null;
+        }
+
+        var fieldName = _fieldNameConverter.ToProviderFieldName(termAggregationRequest.FieldName);
+        var fieldType = schema.Fields.ContainsKey(fieldName) ? (FieldType?)schema.Fields[fieldName] : null;
+
+        ValueFacet result = null;
+
+        switch (fieldType)
+        {
+            case null:
+                break;
+            case FieldType.Text:
+            case FieldType.Number:
+            case FieldType.Date:
+                result = new ValueFacet
+                {
+                    Name = fieldName,
+                    Size = termAggregationRequest.Size == 0 || termAggregationRequest.Size > MaxFacetValues
+                        ? MaxFacetValues
+                        : termAggregationRequest.Size,
+                };
+                break;
+            default:
+                Debug.WriteLine("Elastic App Search supports value facet only for text, number and date fields.");
```

Comment:
Use Logging instead of Debug or otherwise you will not see these messages in Argo/Kubertenes/Docker or on production environments, while they can be important to understand current behavior